bpf/kmesh/ads/include/local_ratelimit.h

```diff
@@ -0,0 +1,197 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Kmesh */
+
+#ifndef __KMESH_LOCAL_RATE_LIMIT_H__
+#define __KMESH_LOCAL_RATE_LIMIT_H__
+
+#include "bpf_log.h"
+#include "bpf_common.h"
+#include "kmesh_common.h"
+#include "listener/listener.pb-c.h"
+
+struct ratelimit_key {
+    union {
+        struct {
+            __u32 ipv4;   /* Destination IPv4 address. */
+            __u32 port;   /* Destination port. */
+            __u32 family; /* Address family (e.g., AF_INET) */
+        } sk_skb;
+    } key;
+};
+
+struct ratelimit_value {
+    __u64 last_topup; /* Timestamp of the last token refill (nanoseconds) */
+    __u64 tokens;     /* Current number of available tokens */
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __type(key, struct ratelimit_key);
+    __type(value, struct ratelimit_value);
+    __uint(max_entries, 1000);
+    __uint(map_flags, BPF_F_NO_PREALLOC);
+} map_of_local_ratelimit SEC(".maps");
+
+struct ratelimit_settings {
+    __u64 bucket_size;       /* Maximum capacity of the token bucket */
+    __u64 tokens_per_topup;  /* Number of tokens added per refill */
+    __u64 topup_interval_ns; /* Interval between token refills (nanoseconds) */
+};
+
+/**
+ * Enforces rate limiting for a given key using token bucket algorithm.
+ *
+ * @param key       Pointer to the rate limit key (e.g., IP and port).
+ * @param settings  Pointer to rate limit settings (bucket size, refill rate, interval).
+ * @return          0 if allowed (token consumed), -1 if rate limit exceeded.
+ */
+static inline int rate_limit__check_and_take(struct ratelimit_key *key, const struct ratelimit_settings *settings)
+{
+    struct ratelimit_value *value;
+    struct ratelimit_value new_value;
+    __u64 now = bpf_ktime_get_ns();
+    __u64 topup;
+
+    value = bpf_map_lookup_elem(&map_of_local_ratelimit, key);
+    if (!value) {
+        new_value.last_topup = now;
+        new_value.tokens = settings->bucket_size;
+        bpf_map_update_elem(&map_of_local_ratelimit, key, &new_value, BPF_ANY);
+        return 0;
+    }
+
+    topup = (now - value->last_topup) / settings->topup_interval_ns;
+    if (topup > 0) {
+        value->tokens += topup * settings->tokens_per_topup;
+        if (value->tokens > settings->bucket_size) {
+            value->tokens = settings->bucket_size;
+        }
+        value->last_topup += topup * settings->topup_interval_ns;
```

Comment:
This may greatly influence the performance.

Reply:
Why would there be a performance impact? Since the interval is typically set to 1 second, the if block runs only about once every second.

Reply:
We could not assume the interval cannot be smaller. In addition, the tricky point is concurrent connections.

Reply:
Or just using [bpf_spin_lock](https://docs.ebpf.io/linux/helper-function/bpf_spin_lock/), but it hurt performance.